### PR TITLE
[flang][Driver] Support -print-supported-cpus and associated aliases

### LINKED ISSUE #117010: [flang] -mcpu=help does not work with flang
Just an odd thing, but on both my Homebrew flang-new (version 19.1.3 on macOS) and on a built-from-git flang on Linux (flang version 20.0.0git (git@github.com:GEOS-ESM/build-llvm-flang.git 697744d14fec6505d07f95249388ac2452ee5151)), if you run `flang -help` you'll see:
```
  -mcpu=<value>           For a list of available CPUs for the target use '-mcpu=help'
```

However, `flang -mcpu=help` does not seem to work on either. On flang 19:
```
❯ flang-new -mcpu=help
ld: library 'c++' not found
flang-new: error: linker command failed with exit code 1 (use -v to see invocation)
```
and on flang 20:
```
> flang -mcpu=help
flang-20: error: no input files
```

So, just in case it needs a file, flang 19:
```
❯ flang-new -c -mcpu=help test.F90
flang-new: error: unsupported argument 'help' to option '-mcpu='
```
and flang 20:
```
> flang -c -mcpu=help test.F90
flang-20: error: unsupported option '-mcpu=' for target 'x86_64-unknown-linux-gnu'
```

Note that `clang -mcpu=help` works on both:
```
❯ clang -mcpu=help |& head
Homebrew clang version 19.1.3
Target: arm64-apple-darwin23.6.0
Thread model: posix
InstalledDir: /Users/mathomp4/.homebrew/brew/Cellar/llvm/19.1.3/bin
Configuration file: /Users/mathomp4/.homebrew/brew/etc/clang/arm64-apple-darwin23.cfg
Available CPUs for this target:

	a64fx
	ampere1
	ampere1a
```
and:
```
> clang -mcpu=help |& head
clang version 20.0.0git (git@github.com:GEOS-ESM/build-llvm-flang.git 697744d14fec6505d07f95249388ac2452ee5151)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /ford1/share/gmao_SIteam/llvm-flang/2024-10-22/bin
Available CPUs for this target:

	alderlake
	amdfam10
	arrowlake
	arrowlake-s
```

### PR DESCRIPTION
**Pull request by tarunprabhu**
The aliases are -mcpu=help and -mtune=help. There is still an issue with the output which prints an example line that references clang. That is not fixed here because it is printed in llvm/MC/SubtargetInfo.cpp. Some more thought is needed to determine how best to handle this.

Fixes #117010

----------------------
Notes for reviewers: The corresponding test in `clang/test/Driver/print-supported-cpus.c` tests two different architectures in the same file. This has been split into two separate tests here to allow the test to be run in cases where only one of the targets has been built.

### CODE REVIEW

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -5899,8 +5899,10 @@ def print_enabled_extensions : Flag<["-", "--"], "print-enabled-extensions">,
   HelpText<"Print the extensions enabled by the given target and -march/-mcpu options."
            " (AArch64 and RISC-V only)">,
   MarshallingInfoFlag<FrontendOpts<"PrintEnabledExtensions">>;
-def : Flag<["-"], "mcpu=help">, Alias<print_supported_cpus>;
-def : Flag<["-"], "mtune=help">, Alias<print_supported_cpus>;
+def : Flag<["-"], "mcpu=help">, Alias<print_supported_cpus>,
+  Visibility<[ClangOption, CC1Option, CLOption, FlangOption, FC1Option]>;
+def : Flag<["-"], "mtune=help">, Alias<print_supported_cpus>,
+  Visibility<[ClangOption, CC1Option, CLOption, FlangOption, FC1Option]>;
```
Comment (banach-space) [inline clang/include/clang/Driver/Options.td:5905]: I'd be tempted to extract these to some documented block, similar to https://github.com/llvm/llvm-project/blob/4086ead63c7e0b56b3b07873117bb7ad7a02d41c/clang/include/clang/Driver/Options.td#L3672-L3691.

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -4417,7 +4417,8 @@ void Driver::BuildActions(Compilation &C, DerivedArgList &Args,
 
       // Use the -mcpu=? flag as the dummy input to cc1.
       Actions.clear();
-      Action *InputAc = C.MakeAction<InputAction>(*A, types::TY_C);
+      Action *InputAc = C.MakeAction<InputAction>(
+          *A, IsFlangMode() ? types::TY_Fortran : types::TY_C);
```
Comment (banach-space) [inline clang/lib/Driver/Driver.cpp:4421]: Unrelated change?

**Comment by banach-space** [comment]
> Notes for reviewers: The corresponding test in clang/test/Driver/print-supported-cpus.c tests two different architectures in the same file. This has been split into two separate tests here to allow the test to be run in cases where only one of the targets has been built.

Wouldn't something like this work: https://github.com/llvm/llvm-project/blob/4086ead63c7e0b56b3b07873117bb7ad7a02d41c/flang/test/Driver/tune-cpu-fir.f90#L1-L3

?

**Review by banach-space — COMMENTED** [review]
Thanks, makes sense. From what I can tell, the behaviour is consistent with Clang (i.e. `-print-supported-cpus` invalidates any other action flags).

Just a couple of smallish comments/suggestions

File: clang/lib/Driver/Driver.cpp (hunk 2 above)
Comment (tarunprabhu) [reply clang/lib/Driver/Driver.cpp:4421]: This is actually needed. 

The `-mcpu=help` argument is used as a "dummy input" (there is a comment a few lines above that says as much). 
The input action created here eventually gets used in `ShouldUseFlangCompiler` which checks if the input type is accepted by `flang`. `flang` does not accept inputs of type `TY_C` (even if they are "dummies"). If `ShouldUseFlangCompiler` returns `false`, a `cc1` command line is created - which is, obviously, not correct.

**Comment by tarunprabhu** [comment]
> > Notes for reviewers: The corresponding test in clang/test/Driver/print-supported-cpus.c tests two different architectures in the same file. This has been split into two separate tests here to allow the test to be run in cases where only one of the targets has been built.
> 
> Wouldn't something like this work:

Ah! I keep forgetting that that is now possible. Updated the tests to do that.

File: clang/include/clang/Driver/Options.td (hunk 1 above)
Comment (tarunprabhu) [reply clang/include/clang/Driver/Options.td]: That's a good idea. I have moved the options into such a block. I have begun adding support for `-print-supported-extensions` and `-print-enabled-extensions` as well which can eventually be added to the block.

File: flang/test/Driver/print-supported-cpus.f90 [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+! X86: Target: x86_64-unknown-linux-gnu
+! X86: corei7
+
+! AARCH64: Target: aarch64-unknown-linux-gnu
+! AARCH64: cortex-a73
+! AARCH64: cortex-a75
+
+! TODO: This is a line that is printed at the end of the output. The full line
+! also includes an example that references clang. That needs to be fixed and a
+! a check added here to make sure that it references flang, not clang.
```
Comment (banach-space) [inline flang/test/Driver/print-supported-cpus.f90:41]: [nit] I would add that bit with clang invocation. Once this is fixed, the test will fail and we will know that we don't need to write a new test (sometimes less is more)

**Review by banach-space — APPROVED** [review]
LGTM, my final comment is nit - feel free to ignore. Thanks!

File: flang/test/Driver/print-supported-cpus.f90 (hunk 3 above)
Comment (tarunprabhu) [reply flang/test/Driver/print-supported-cpus.f90]: That makes sense.  I added that to the test and updated the comment. It will fail when `llvm/MCSubtargetInfo` is fixed.

**Review by MaskRay — APPROVED** [review]

**Comment by jplehr** [comment]
I believe this results in tests failures on our bot: https://lab.llvm.org/staging/#/builders/105/builds/9796 (and a few others)
